### ISSUE TITLE
failed dkms build results only in warning

### ISSUE THREAD

**Issue by btwe** [actionless/pikaur#591]
```
% pikaur -Vq

Pikaur v1.8
Pacman v6.0.1 - libalpm v13.0.1 - pyalpm v0.10.6
```

##### Description:
This issue was triggered by a failed dkms build. I frequently run `pikaur --noconfirm -Suy && backup && reboot`.
In my case, the automatically triggered build for zfs-dkms failed, but pikaur only printed a WARNING, but eventually returned SUCCESS. Is this suppose to happen?

This rendered my system unusable after reboot. So my wish would be to let pikaur return an error code, if something along the upgrade fails.

##### Attached log:
snap:
```
(4/5) Install DKMS modules
==> dkms install --no-depmod -m zfs -v 2.1.0 -k 5.14.2-arch1-2
Error! Bad return status for module build on kernel: 5.14.2-arch1-2 (x86_64)
Consult /var/lib/dkms/zfs/2.1.0/build/make.log for more information.
==> Warning, `dkms install --no-depmod -m zfs -v 2.1.0 -k 5.14.2-arch1-2' returned 10
==> depmod 5.14.2-arch1-2
```

full:
```
% pikaur -S --noconfirm zfs-dkms --verbose --pikaur-debug
=> systemd-run --service-type=oneshot --pipe --wait --pty -p DynamicUser=yes -p CacheDirectory=pikaur -E HOME=/tmp true
:: debug: Pikaur operation found for sys.argv=['/usr/bin/pikaur', '-S', '--noconfirm', 'zfs-dkms', '--verbose', '--pikaur-debug']: cli_in
stall_packages
:: debug: install_info_fetcher: Gonna fetch install info for:
    install_package_names=['zfs-dkms']
    not_found_repo_pkgs_names=[]
    pkgbuilds_packagelists={}
    manually_excluded_packages_names=[]
:: debug: install_info_fetcher: Gonna get repo pkgs install info...
Reading repository package databases...
Reading local package database...
=> pacman --color=never --sync --noconfirm zfs-dkms --print-format %r/%n
:: debug: install_info_fetcher: gonna get AUR pkgs install info for:
    aur_packages_versionmatchers=[]
    self.aur_updates_install_info=[]
    aur_packages_names_to_versions={}
:: debug: install_info_fetcher: got AUR pkgs install info: self.aur_updates_install_info=[]

=> pacman --color=never --sync --verbose --noconfirm zfs-dkms
Root      : /
Conf File : /etc/pacman.conf
DB Path   : /var/lib/pacman/
Cache Dirs: /var/cache/pacman/pkg/
Hook Dirs : /usr/share/libalpm/hooks/  /etc/pacman.d/hooks/
Lock File : /var/lib/pacman/db.lck
Log File  : /var/log/pacman.log
GPG Dir   : /etc/pacman.d/gnupg/
Targets   : zfs-dkms
resolving dependencies...
looking for conflicting packages...

Packages (2) zfs-utils-2.1.0-2  zfs-dkms-2.1.0-1

Total Installed Size:  103.98 MiB

:: Proceed with installation? [Y/n]
checking keyring...
checking package integrity...
loading package files...
checking for file conflicts...
checking available disk space...
:: Processing package changes...
installing zfs-utils...
Optional dependencies for zfs-utils
    python: pyzfs and extra utilities, [installed]
    python-cffi: pyzfs [installed]
installing zfs-dkms...
:: Running post-transaction hooks...
(1/5) Reloading system manager configuration...
(2/5) Reloading device manager configuration...
(3/5) Arming ConditionNeedsUpdate...
(4/5) Install DKMS modules
==> dkms install --no-depmod -m zfs -v 2.1.0 -k 5.14.2-arch1-2
Error! Bad return status for module build on kernel: 5.14.2-arch1-2 (x86_64)
Consult /var/lib/dkms/zfs/2.1.0/build/make.log for more information.
==> Warning, `dkms install --no-depmod -m zfs -v 2.1.0 -k 5.14.2-arch1-2' returned 10
==> depmod 5.14.2-arch1-2
(5/5) Updating linux initcpios...
==> Building image from preset: /etc/mkinitcpio.d/linux.preset: 'default'
  -> -k /boot/vmlinuz-linux -c /etc/mkinitcpio.conf -g /boot/initramfs-linux.img
==> Starting build: 5.14.2-arch1-2
  -> Running build hook: [base]
  -> Running build hook: [udev]
  -> Running build hook: [autodetect]
  -> Running build hook: [modconf]
  -> Running build hook: [block]
==> WARNING: Possibly missing firmware for module: xhci_pci
  -> Running build hook: [filesystems]
  -> Running build hook: [keyboard]
  -> Running build hook: [fsck]
==> Generating module dependencies
==> Creating zstd-compressed initcpio image: /boot/initramfs-linux.img
==> Image generation successful
==> Building image from preset: /etc/mkinitcpio.d/linux.preset: 'fallback'
  -> -k /boot/vmlinuz-linux -c /etc/mkinitcpio.conf -g /boot/initramfs-linux-fallback.img -S autodetect
==> Starting build: 5.14.2-arch1-2
  -> Running build hook: [base]
  -> Running build hook: [udev]
  -> Running build hook: [modconf]
  -> Running build hook: [block]
==> WARNING: Possibly missing firmware for module: aic94xx
==> WARNING: Possibly missing firmware for module: wd719x
==> WARNING: Possibly missing firmware for module: xhci_pci
  -> Running build hook: [filesystems]
  -> Running build hook: [keyboard]
  -> Running build hook: [fsck]
==> Generating module dependencies
==> Creating zstd-compressed initcpio image: /boot/initramfs-linux-fallback.img
==> Image generation successful
pikaur -S --noconfirm zfs-dkms --verbose --pikaur-debug  633.16s user 64.88s system 554% cpu 2:05.80 total
tee /tmp/pikaur-dksm-error.log  0.00s user 0.01s system 0% cpu 2:05.74 total

% echo $?
0

```

**Comment by actionless** [comment]
are you sure what the same pacman command returns any different return code? (`pacman --color=never --sync --verbose --noconfirm zfs-dkms`)

**Comment by btwe** [comment]
Thanks for that pointer. You are right:
```
root@zam709 ~ # pacman --color=never --sync --verbose --noconfirm zfs-dkms; echo $?
Root      : /
Conf File : /etc/pacman.conf
DB Path   : /var/lib/pacman/
Cache Dirs: /var/cache/pacman/pkg/
Hook Dirs : /usr/share/libalpm/hooks/  /etc/pacman.d/hooks/
Lock File : /var/lib/pacman/db.lck
Log File  : /var/log/pacman.log
GPG Dir   : /etc/pacman.d/gnupg/
Targets   : zfs-dkms
resolving dependencies...
looking for conflicting packages...

Packages (2) zfs-utils-2.1.0-2  zfs-dkms-2.1.0-1

Total Download Size:    29.03 MiB
Total Installed Size:  103.98 MiB

:: Proceed with installation? [Y/n]
:: Retrieving packages...
 zfs-utils-2.1.0-2-x86_64                               29.0 MiB  22.4 MiB/s 00:01 [###############################################] 100%
(2/2) checking keys in keyring                                                     [###############################################] 100%
(2/2) checking package integrity                                                   [###############################################] 100%
(2/2) loading package files                                                        [###############################################] 100%
(2/2) checking for file conflicts                                                  [###############################################] 100%
(2/2) checking available disk space                                                [###############################################] 100%
:: Processing package changes...
(1/2) installing zfs-utils                                                         [###############################################] 100%
Optional dependencies for zfs-utils
    python: pyzfs and extra utilities, [installed]
    python-cffi: pyzfs [installed]
(2/2) installing zfs-dkms                                                          [###############################################] 100%
:: Running post-transaction hooks...
(1/5) Reloading system manager configuration...
(2/5) Reloading device manager configuration...
(3/5) Arming ConditionNeedsUpdate...
(4/5) Install DKMS modules
==> dkms install --no-depmod -m zfs -v 2.1.0 -k 5.14.2-arch1-2
Error! Bad return status for module build on kernel: 5.14.2-arch1-2 (x86_64)
Consult /var/lib/dkms/zfs/2.1.0/build/make.log for more information.
==> Warning, `dkms install --no-depmod -m zfs -v 2.1.0 -k 5.14.2-arch1-2' returned 10
==> depmod 5.14.2-arch1-2
(5/5) Updating linux initcpios...
==> Building image from preset: /etc/mkinitcpio.d/linux.preset: 'default'
  -> -k /boot/vmlinuz-linux -c /etc/mkinitcpio.conf -g /boot/initramfs-linux.img
==> Starting build: 5.14.2-arch1-2
  -> Running build hook: [base]
  -> Running build hook: [udev]
  -> Running build hook: [autodetect]
  -> Running build hook: [modconf]
  -> Running build hook: [block]
==> WARNING: Possibly missing firmware for module: xhci_pci
  -> Running build hook: [filesystems]
  -> Running build hook: [keyboard]
  -> Running build hook: [fsck]
==> Generating module dependencies
==> Creating zstd-compressed initcpio image: /boot/initramfs-linux.img
==> Image generation successful
==> Building image from preset: /etc/mkinitcpio.d/linux.preset: 'fallback'
  -> -k /boot/vmlinuz-linux -c /etc/mkinitcpio.conf -g /boot/initramfs-linux-fallback.img -S autodetect
==> Starting build: 5.14.2-arch1-2
  -> Running build hook: [base]
  -> Running build hook: [udev]
  -> Running build hook: [modconf]
  -> Running build hook: [block]
==> WARNING: Possibly missing firmware for module: wd719x
==> WARNING: Possibly missing firmware for module: aic94xx
==> WARNING: Possibly missing firmware for module: xhci_pci
  -> Running build hook: [filesystems]
  -> Running build hook: [keyboard]
  -> Running build hook: [fsck]
==> Generating module dependencies
==> Creating zstd-compressed initcpio image: /boot/initramfs-linux-fallback.img
==> Image generation successful
pacman --color=never --sync --verbose --noconfirm zfs-dkms  732.76s user 81.43s system 306% cpu 4:25.30 total
0
```

Also pacman exits with SUCCESS. So this one could be closed from pikaur pov. 

(I exclusively use pikaur since I started with arch, so kudos and many thanks to you)

**Comment by thurstylark** [comment]
@btwe I clicked through from your Arch bug report (https://bugs.archlinux.org/task/72130) which has already been closed with the reason "working as intended", and since comments are closed there, I wanted to give a tad bit of context.

According to the [`alpm-hooks`](https://man.archlinux.org/man/alpm-hooks.5#ACTIONS) manpage, hooks can use the `AbortOnFail` option to abort the transaction, but that's only available for PreTransaction hooks. That's because it would be a bad idea to allow an arbitrary hook to abort any PostTransaction hooks that have yet to run their own important cleanup procedures. As far as pacman is aware, the transaction completed successfully. *dkms* is what failed. It's not pacman's job to exit with a nonzero code when a hook fails.

Additionally, [pacman is not intended to be run blindly](https://wiki.archlinux.org/title/Pacman#Upgrading_packages). It is the user's job to read pacman's output and act on what it says if necessary. The problem you're having is the direct result of not following this guidance, and that's why your bug was summarily closed.
